### PR TITLE
PutBlob, add one option: skipCompositeChunk.

### PR DESCRIPTION
**Pull request by JingQianCloud**
PutBlob, we add one option: skipCompositeChunk.
When it's true, we don't generate the composite blob.
And then we return serialized PutBlobMetaInfo string instead of the composite blob id.

### CODE REVIEW

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java [hunk 1]
```diff
@@ -2077,6 +2084,10 @@ void poll(RequestRegistrationCallback<PutOperation> requestRegistrationCallback)
       }
     }
 
+    private boolean isS3MultipartPartUpload() {
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:2087]: When we have the MultiPart code, we may set a flag called "Skip-Composite-Blob" flag.
So here we rely on this flag to skip the compositeBlob step.

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java [hunk 2]
```diff
@@ -2089,7 +2100,34 @@ private void finalizeMetadataChunk() {
...
+          jsonObject.accumulate("chunks", oneObject);
+        }
+        jsonObject.put("ttl", ""); // placeholder
+        jsonObject.put("reservedMetadataId", ""); // placeholder
+        jsonObject.put("session", ""); // placeholder
+        String result = jsonObject.toString();
+        logger.info("S3 5MB : finalizeMetadataChunk generated json {}", result);
+
+        // use the S3_BLOB_LIST to pass the value
+        restRequest.setArg(S3_BLOB_LIST, result);
```
Comment (JingQianCloud) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:2129]: Ideally we let S3 code path to generate the ETag, not here. 
Anyway to pass the dataChunk raw data to the caller?

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java [hunk 3]
```diff
@@ -2089,7 +2100,34 @@ private void finalizeMetadataChunk() {
               passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
               passedInBlobProperties.isEncrypted(), passedInBlobProperties.getExternalAssetTag(),
               passedInBlobProperties.getContentEncoding(), passedInBlobProperties.getFilename(), null);
-      if (isStitchOperation() || getNumDataChunks() > 1) {
+
+      if (isS3MultipartPartUpload()) {
```
Comment (justinlin-linkedin) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:2104]: I don't think this is a good idea to have s3 knowledge inside of PutOperation. I would propose another way of convey these information in and out. Basically, we have extra input (a flag to indicate if we want to have composite blob id) and extra output (a list of blob ids in a string). I think we can use PutBlobOptions to pass this information, something PutBlobOption.ignoreCompositeBlobId, and just return a string (maybe in base64) that contains all blob ids and sizes. Something like that.

**Review by justinlin-linkedin — APPROVED** [review]
LGTM

File: ambry-router/src/main/java/com/github/ambry/router/PutBlobMetaInfo.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+   */
+  public int getChunkNumber() {
+    return orderedChunkIdSizeList.size();
+  }
+
+  /**
+   * Get the chunk list
+   * @return the list of chunk with blob id and blob size
+   */
+  public List<Pair<String, Long>> getOrderedChunkIdSizeList() {
```
Comment (weisong44) [inline ambry-router/src/main/java/com/github/ambry/router/PutBlobMetaInfo.java:75]: Nit: would it make sense to make the returned list immutable?

**Review by weisong44 — COMMENTED** [review]
I don't think I know enough to do a proper review, just left a very minor comment :slightly_smiling_face:

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java [hunk 5]
```diff
@@ -2089,7 +2101,25 @@ private void finalizeMetadataChunk() {
               passedInBlobProperties.getAccountId(), passedInBlobProperties.getContainerId(),
               passedInBlobProperties.isEncrypted(), passedInBlobProperties.getExternalAssetTag(),
               passedInBlobProperties.getContentEncoding(), passedInBlobProperties.getFilename(), null);
-      if (isStitchOperation() || getNumDataChunks() > 1) {
+
+      if (options.skipCompositeChunk()) {
+        // close the request as the single blob. we don't generate the composite blob.
+        blobId = (BlobId) indexToChunkIdsAndChunkSizes.get(0).getFirst();
```
Comment (Arun-LinkedIn) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:2107]: Setting blobId to the first chunk ID looks bit hacky :) But, I also don't know any better way.

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java [hunk 6]
```diff
@@ -2089,7 +2101,25 @@ private void finalizeMetadataChunk() {
...
+        // now generate the PutBlobMetaInfo which includes the data chunk list
+        // values returned are in the right order as TreeMap returns them in key-order.
+        // .stream.map.collect is supposed to return in order
+        List<Pair<String, Long>> orderedChunkList = new ArrayList<>(indexToChunkIdsAndChunkSizes.values()
+            .stream()
+            .map(p -> new Pair<>(p.getFirst().toString(), p.getSecond()))
+            .collect(Collectors.toList()));
+        PutBlobMetaInfo putBlobMetaInfoObj = new PutBlobMetaInfo(orderedChunkList, reservedMetadataChunkId.getID());
+        putBlobMetaInfo = PutBlobMetaInfo.serialize(putBlobMetaInfoObj);
+        logger.info("S3 5MB : finalizeMetadataChunk orderedChunkIdList is {} {}", indexToChunkIdsAndChunkSizes,
```
Comment (Arun-LinkedIn) [inline ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:2120]: nit: Would `debug` level be sufficient?

**Review by Arun-LinkedIn — APPROVED** [review]
Looks good to me. Just had couple of nit comments.

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java (hunk 5 above)
Comment (JingQianCloud) [reply ambry-router/src/main/java/com/github/ambry/router/PutOperation.java:2108]: That's true. Added comment to explain it. Hopefully it's less confusing. :-)

File: ambry-router/src/main/java/com/github/ambry/router/PutOperation.java (hunk 6 above)
Comment (JingQianCloud) [reply ambry-router/src/main/java/com/github/ambry/router/PutOperation.java]: Changed.

**Pull request merged**